File: src/main/java/链表/两数相加.java
package 链表;

import type.ListNode;

public class 两数相加 {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        if (l1.next == null && l1.val == 0) {
            return l2;
        }
        if (l2.next == null && l2.val == 0) {
            return l1;
        }
        ListNode s1 = l1;
        ListNode s2 = l2;
        ListNode result = new ListNode(0);
        ListNode r = result;
        int carry = 0;

        while (s1 != null || s2 != null) {
            if (s1 == null) {
                r.next = new ListNode((carry + s2.val) % 10);
                carry = (carry + s2.val) / 10;
                s2 = s2.next;
                r = r.next;
                continue;
            }
            if (s2 == null) {
                r.next = new ListNode((carry + s1.val) % 10);
                carry = (carry + s1.val) / 10;
                s1 = s1.next;
                r = r.next;
                continue;
            }
            r.next = new ListNode((carry + s1.val + s2.val) % 10);
            carry = (carry + s1.val + s2.val) / 10;
            s1 = s1.next;
            s2 = s2.next;
            r = r.next;
        }
        if (carry > 0) {
            r.next = new ListNode(1);
        }
        return result.next;
    }

    public static void main(String[] args) {
        两数相加 test = new 两数相加();
        ListNode l1 = new ListNode(2);
        l1.next = new ListNode(4);
        l1.next.next = new ListNode(3);
        ListNode l2 = new ListNode(5);
        l2.next = new ListNode(6);
        l2.next.next = new ListNode(4);
        ListNode listNode = test.addTwoNumbers(l1, l2);
        System.out.println(listNode);
    }
}

File: src/main/java/done/两数相加.kt
//给出两个 非空 的链表用来表示两个非负的整数。其中，它们各自的位数是按照 逆序 的方式存储的，并且它们的每个节点只能存储 一位 数字。
//
// 如果，我们将这两个数相加起来，则会返回一个新的链表来表示它们的和。
//
// 您可以假设除了数字 0 之外，这两个数都不会以 0 开头。
//
// 示例：
//
// 输入：(2 -> 4 -> 3) + (5 -> 6 -> 4)
//输出：7 -> 0 -> 8
//原因：342 + 465 = 807
//
// Related Topics 链表 数学

/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class 两数相加 {
    class ListNode(var `val`: Int) {
        var next: ListNode? = null
    }

    fun addTwoNumbers(l1: ListNode?, l2: ListNode?): ListNode? {
        var carry = 0
        val result = ListNode(0)
        var l1c: ListNode? = l1
        var l2c: ListNode? = l2
        var c: ListNode? = result

        while (l1c != null || l2c != null) {
            var value = (l1c?.`val` ?: 0) + (l2c?.`val` ?: 0) + carry
            if (value > 9) {
                value %= 10
                carry = 1
            } else {
                carry = 0
            }
            c?.next = ListNode(value)
            c = c?.next
            if (l1?.next != null) l1c = l1c?.next
            if (l2?.next != null) l2c = l2c?.next
        }
        if (carry == 1) {
            c?.next = ListNode(1)
        }
        return result.next
    }
}